src/dof_management/assemble_node_freedom_allocation_table.hpp

```diff
@@ -5,23 +5,57 @@
 #include "freedom_signature.hpp"
 
 #include "src/beams/beams.hpp"
+#include "src/constraints/constraints.hpp"
 #include "src/state/state.hpp"
 
 namespace openturbine {
 
-inline void assemble_node_freedom_allocation_table(State& state, const Beams& beams) {
+inline void assemble_node_freedom_allocation_table(
```

Comment:
That would be an option.  3 thoughts:

1.  But then we would need to add this logic to each the `BeamNode`, `Constraint`, and whatever high level equivalents that we come up with for future element types.  Basically, we'd be requiring each of our high level classes to know how to combine an element signature.

2.  While we don't have the ability in there right now, what if you want to remove an element from the model?  To get the right signature, you'd basically need to rebuild the whole model, node-by-node, from scratch (or else have special logic kind of like this but walking the different elements in our model).  

3.  We'll have to get our mapping information onto the GPU at some point for constructing and working with our sparse matrices, so it makes sense to go ahead and construct it there rather than doing it on CPU in an object oriented manner and then copying it up.  

I considered having different overloads for each `Beams` and `Constraints` (and future element types) which each do the action of walking those structures.  Then there would be no single function that would need to know the internals of each and every element type.  This would work since the `|` operator commutes.  I'm not against making this change in the future, and will likely implicitly have that structure once I parallelize this algorithm.

Reply:
Sounds good, thanks